File: Review-02/Example-5/source/House.cpp
#include "House.h"

/**
 *
 */
House::House()
    :name("House"),
     rooms()
{
}

/**
 *
 */
House::House(std::string name)
    :name(name)
{
    //this->name = name;
}

/**
 *
 */
void House::addRoom(Room toAdd)
{
    rooms.push_back(toAdd);
}

/**
 *
 */
size_t House::size() const {
    return rooms.size();
}

/**
 *
 */
House::iterator House::begin()
{
    return rooms.begin();
}

/**
 *
 */
House::const_iterator House::begin() const
{
    return rooms.begin();
}

/**
 *
 */
House::iterator House::end()
{
    return rooms.end();
}

/**
 *
 */
House::const_iterator House::end() const
{
    return rooms.end();
}

/**
 *
 */
void House::display(std::ostream& outs) const
{
    outs << "--------" << this->name << "--------" << "\n";

    // What type of iterator am I using--i.e.,
    // iterator or const_iterator?

    // Print the rooms
    for (const Room& prtRoom : *this) {
        outs << prtRoom;
    }

    // Compute and print the total
    double                avg   = 0;
    double                total = 0;

    for (const Room& room : *this) {
        total += room.flooringCost();
    }

    avg = total / this->size();

    outs << "\n";
    outs << "------------------------------";
    outs << "\n";

    outs << "Total Cost   : $ " << total << "\n";
    outs << "Avg Room Cost: $ " << avg   << "\n";
}

File: Review-02-Java-Well-Defined-Interfaces/Example-2/src/main/java/house/House.java
package house;

import java.util.Collection;
import java.util.Iterator;
import java.util.ArrayList;

/**
 * A House is composed of zero or more Room objects.
 * <p>
 * This class serves as our demonstration of the
 * iterator interface.
 */
public class House implements Iterable<Room>
{
    /**
     * Name of the house--e.g.,
     * Minecraft Beach House
     */
    private String name;

    // private LinkedList<Room> rooms;
    private Collection<Room> rooms;

    /**
     * Construct a House with a
     * generic name and no rooms.
     */
    public House()
    {
        this.name = "Generic";
        this.rooms = new ArrayList<>();
    }

    /**
     * Construct a House with a specified name
     */
    public House(String name)
    {
        this.name = name;
        this.rooms = new ArrayList<>();
    }

    /**
     * Add a Room
     *
     * @param toAdd new Room object
     */
    public void addRoom(Room toAdd)
    {
        this.rooms.add(toAdd);
    }

    /**
     * Update the name
     */
    public void setName(String newName)
    {
        this.name = newName;
    }

    /**
     * Update the name
     */
    public String getName()
    {
        return this.name;
    }

    /**
     * Allow access to the _beginning_ of the
     * house--i.e., Room container--via an
     * iterator.
     */
    public Iterator<Room> iterator()
    {
        return this.rooms.iterator();
    }

    /**
     * Return the size of the house--i.e.,
     * the number of rooms
     */
    public int size()
    {
        return this.rooms.size();
    }

    /**
     * Logical Equivalence Operator
     */
    @Override
    public boolean equals(Object rhsObj)
    {
        if (!(rhsObj instanceof House)) {
            return false;
        }

        House rhs = (House) rhsObj;

        if (!this.name.equals(rhs.name)) {
            return false;
        }

        if (this.size() != rhs.size()) {
           return false;
        }

        Iterator<Room> lhsIt = this.iterator();
        Iterator<Room> rhsIt = rhs.iterator();

        // We know that both houses have the same number of rooms.
        while (lhsIt.hasNext()) {
            Room lhsRoom = lhsIt.next();
            Room rhsRoom = rhsIt.next();

            if (!lhsRoom.equals(rhsRoom)) {
                return false;
            }
        }

        return true;
    }

    @Override
    public String toString()
    {
        // I will use String concatenation in this example. We will discuss the
        // **proper** and efficient** way to combine strings in Java in a future
        // example.

        String toOutput = String.format("--------%s--------%n", this.name);

        double totalCost = 0;
        for (Room rm : this) {
            toOutput += rm + String.format("%n");

            totalCost += rm.flooringCost();
        }

        double avg   = totalCost / this.size();

        toOutput += String.format("------------------------------%n");
        toOutput += String.format("Total Cost   : $ %.2f%n", totalCost);
        toOutput += String.format("Avg Room Cost: $ %.2f", avg);

        return toOutput;
    }
}
